### PR TITLE
feat: add scale functionality for stream

### PR DESCRIPTION
**Pull request by sergiupantiru**
Added functionality to scale applications for a stream.

Screenshots:
image
image
image
image

Signed-off-by: Sergiu Pantiru <sergiu.pantiru@gmail.com>

### CODE REVIEW

**Review by claudiahub — COMMENTED** [review]
Hi Sergiu,

Thanks for your contribution!
Great to see tests as well.

Code looks good, a couple of items should be adjusted.

1. `ui/src/app/streams/streams/scale/scale.component.ts` line 62: class Instance should be moved into separated file and imported correspondingly.

2. `ui/src/app/streams/streams/scale/scale.component.html` from line 12 to line 46: input field and submit button should be wrapped into a form and validated. Please note we do not consider scaling down to zero or negative number of instances

3. `ui/src/app/streams/streams/scale/scale.component.html` lines 16-17 from inline style and generated output css: On the Scale instances for app table, padding seems different from our platform tables. 
Please see my below screenshots

a) Platform table with cta action button
12px 18px

 
b) Scale instances table, see padding values
Streams

If you need further information, feel free to contact us,
we appreciate your contribution.

Thanks!

Claudia

**Comment by sergiupantiru** [comment]
Thanks Claudia for looking into this. For the table I think I took as an example the Managed Versions table.
image

**Comment by sergiupantiru** [comment]
@claudiahub I did the validation (not via form group) and it looks like this.

image

Please let me know if this is ok and if not I will try to do it via form group.

Also I think the CLI allows to scale to 0. Should the UI and CLI not be aligned?

I will look also at the padding problem.

**Comment by claudiahub** [comment]
@sergiupantiru Thanks,

For the padding, I've checked with the team and the styles you did already for the table looks good.

**Comment by sergiupantiru** [comment]
Gave it a try with FormGroup. The validation is for min 0 without a max. Please let me know if this is not ok.

File: ui/src/app/streams/streams/scale/scale.component.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  initForm(metrics: StreamStatus[]): void {
+    this.instances = [];
+    if (metrics && metrics.length > 0) {
+      metrics[0].applications.forEach(appStatus => {
+        const instance = Instance.fromAppStatus(appStatus);
+        this.instances.push(instance);
+        this.form.addControl(
+          `instance${instance.name}`,
+          new FormControl(instance.instanceCount, [Validators.required, Validators.min(0), Validators.max(9)])
```
Comment (claudiahub) [inline ui/src/app/streams/streams/scale/scale.component.ts]: Instances minimum value should be 1

File: ui/src/app/streams/streams/scale/scale.component.html [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          <ng-container>
+            <tr *ngFor="let instance of instances; index as i">
+              <td class="left middle">
+                {{ instance.name }}
+              </td>
+              <td class="left">
+                <input
+                  style="width: 50px"
+                  type="number"
+                  min="0"
```
Comment (claudiahub) [inline ui/src/app/streams/streams/scale/scale.component.html]: Instances minimum value should be 1

**Review by claudiahub — APPROVED** [review]

**Review by oodamien — APPROVED** [review]
LGTM, thanks @sergiupantiru for the contribution

**Comment by sergiupantiru** [comment]
Thank you both for the help with my pr. 
May I ask why should it be limited to maximum 9 the scale? If we deploy to kubernetes we should be able to scale much more. (Me personally I've tested with 20 already, via CLI).

**Comment by oodamien** [comment]
@sergiupantiru It will be updated before release.
@jvalkeal What do you think?

**Pull request merged**